grype/pkg/package_test.go

```diff
@@ -198,375 +227,14 @@ func TestNew(t *testing.T) {
-					Name:            "a",
-					Architecture:    "a",
-					Version:         "a",
-					ExtendedVersion: "a",
-					BuildTime:       "a",
-					Author:          "a",
-					Format:          "a",
-					RWRootFS:        true,
-					SwapDevice:      10,
-					RootDevice:      11,
-					VideoMode:       "a",
-				},
-			},
-		},
-		{
-			name: "linux-kernel-module-metadata",
-			syftPkg: syftPkg.Package{
-				MetadataType: syftPkg.LinuxKernelModuleMetadataType,
-				Metadata: syftPkg.LinuxKernelModuleMetadata{
-					Name:          "a",
-					Version:       "a",
-					SourceVersion: "a",
-					Path:          "a",
-					Description:   "a",
-					Author:        "a",
-					License:       "a",
-					KernelVersion: "a",
-					VersionMagic:  "a",
-					Parameters: map[string]syftPkg.LinuxKernelModuleParameter{
-						"a": {
-							Type:        "a",
-							Description: "a",
-						},
-					},
-				},
-			},
-		},
-		{
-			name: "r-description-file-metadata",
-			syftPkg: syftPkg.Package{
-				MetadataType: syftPkg.RDescriptionFileMetadataType,
-				Metadata: syftPkg.RDescriptionFileMetadata{
-					Title:            "a",
-					Description:      "a",
-					Author:           "a",
-					Maintainer:       "a",
-					URL:              []string{"a"},
-					Repository:       "a",
-					Built:            "a",
-					NeedsCompilation: true,
-					Imports:          []string{"a"},
-					Depends:          []string{"a"},
-					Suggests:         []string{"a"},
-				},
-			},
-		},
-		{
-			name: "dotnet-portable-executable-metadata",
-			syftPkg: syftPkg.Package{
-				MetadataType: syftPkg.DotnetPortableExecutableMetadataType,
-				Metadata: syftPkg.DotnetPortableExecutableMetadata{
-					AssemblyVersion: "a",
-					LegalCopyright:  "a",
-					Comments:        "a",
-					InternalName:    "a",
-					CompanyName:     "a",
-					ProductName:     "a",
-					ProductVersion:  "a",
-				},
-			},
-		},
-		{
-			name: "dotnet-portable-executable-metadata",
-			syftPkg: syftPkg.Package{
-				MetadataType: syftPkg.SwiftPackageManagerMetadataType,
-				Metadata: syftPkg.SwiftPackageManagerMetadata{
-					Revision: "a",
-				},
-			},
-		},
 	}
 
-	// capture each observed metadata type, we should see all of them relate to what syft provides by the end of testing
-	expectedMetadataTypes := set.NewStringSet()
-	for _, ty := range syftPkg.AllMetadataTypes {
-		expectedMetadataTypes.Add(string(ty))
-	}
-
-	// run all of our cases
-	observedMetadataTypes := set.NewStringSet()
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			if string(test.syftPkg.MetadataType) != "" {
-				observedMetadataTypes.Add(string(test.syftPkg.MetadataType))
-			}
 			assert.Equal(t, test.metadata, New(test.syftPkg).Metadata, "unexpected metadata")
 			assert.Equal(t, test.upstreams, New(test.syftPkg).Upstreams, "unexpected upstream")
 		})
 	}
-
-	// did we see all possible metadata types? if not, then there is an uncovered case and this test should error out
```

Comment:
Is it asserted elsewhere that we cover all metadata types? Maybe this logic moved into Syft and I don't know about it? Or maybe I'm missing something and our mechanism changed so that it's no longer necessary to assert completeness here?

Reply:
I should have made a comment in the description about this removal; essentially the fact that we were extracting upstream packages from the metadata was the important part of this test. We were following a completeness testing pattern here to ensure that as new metadata types are added that there is a test that shows what the new behavior should be. Since the metadata type definitions have been moved from a collection of strings to a collection of structs that get a label, most of this has been moved to an `internal` package within syft to keep the API a little more defensible. That also makes completion testing here intractable, since there isn't a single source of truth to get the full collection of expected types to process. For that reason I reduced the scope of this test to simply ensure we are covering known types correctly (only a few types support the upstream packages concept).